File: src/main/java/pl/polsl/cargoflow/model/Employee.java
package pl.polsl.cargoflow.model;

import java.sql.Date;
import java.util.List;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OneToOne;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import pl.polsl.cargoflow.model.dto.EmployeeRequest;

@Entity
@NoArgsConstructor
@Getter
@Setter
public class Employee {
    @Id
    @GeneratedValue(strategy=GenerationType.AUTO)
    private Long id;

    @NotBlank
    private String name;

    @NotBlank
    private String surname;

    @NotBlank
    @Size(min = 11, max = 11)
    private String pesel;

    private Date birthDate;

    @OneToOne(cascade = CascadeType.ALL)
    private Credentials credentials;

    @ManyToOne
    @NotNull
    private Position position;

    @ManyToMany
    private List<DrivingLicense> driverLicenses;

    @OneToMany(mappedBy = "employee", fetch = FetchType.LAZY)
    private List<DriverRoute> driverRoutes;

    public Employee(EmployeeRequest employeeRequest, Position position, List<DrivingLicense> drivingLicenses) {
        this.name = employeeRequest.getName();
        this.surname = employeeRequest.getSurname();
        this.pesel = employeeRequest.getPesel();
        this.birthDate = employeeRequest.getBirthDate();
        this.credentials = new Credentials(employeeRequest.getLogin(), employeeRequest.getPassword());
        this.position = position;
        this.driverLicenses = drivingLicenses;
    }
}

File: web/src/types/Employee.ts
export type PositionName = "admin" | "coordinator" | "driver" | "mechanic";

export interface Position {
    id: number;
    name: PositionName;
}

export interface DriverLicence {
    id: number;
    name: string;
}

export interface User {
    id: number;
    name: string;
    surname: string;
    pesel: string;
    birthDate: string;
    position: Position;
    driverLicenses: DriverLicence[];
}

export type Admin = User & { position: { id: number; name: "admin" } };
export type Coordinator = User & { position: { id: number; name: "coordinator" } };
export type Mechanic = User & { position: { id: number; name: "mechanic" } };
export type Driver = User & {
    position: { id: number; name: "driver" };
};

export type Employee = Admin | Coordinator | Mechanic | Driver;

export function isDriver(employee: Employee): employee is Driver {
    return employee.position.name === "driver";
}